java/org/apache/catalina/filters/CsrfPreventionFilter.java

```diff
@@ -87,11 +104,170 @@ public void setNonceRequestParameterName(String parameterName) {
         this.nonceRequestParameterName = parameterName;
     }
 
+    /**
+     * Sets the flag to enforce CSRF protection or just log failures as DEBUG
+     * messages.
+     *
+     * @param enforce <code>true</code> to enforce CSRF protections or
+     *                <code>false</code> to log DEBUG messages and allow
+     *                all requests.
+     */
+    public void setEnforce(boolean enforce) {
+        this.enforce = enforce;
+    }
+
+    /**
+     * Gets the flag to enforce CSRF protection or just log failures as DEBUG
+     * messages.
+     *
+     * @return <code>true</code> if CSRF protections will be enforced or
+     *         <code>false</code> if all requests will be allowed and
+     *         failures will be logged as DEBUG messages.
+     */
+    public boolean getEnforce() {
+        return this.enforce;
+    }
+
+    /**
+     * Sets the list of URL patterns to suppress nonce-addition for.
+     *
+     * Some URLs do not need nonces added to them such as static resources.
+     * By <i>not</i> adding nonces to those URLs, HTTP caches can be more
+     * effective because the CSRF prevention filter won't generate what
+     * look like unique URLs for those commonly-reused resources.
+     *
+     * @param patterns A comma-separated list of URL patterns that will not
+     *        have nonces added to them. Patterns may begin or end with a
+     *        <code>*</code> character to denote a suffix-match or
+     *        prefix-match. Any matched URL will not have a CSRF nonce
+     *        added to it when passed through
+     *        {@link HttpServletResponse#encodeURL(String)}.
+     */
+    public void setNoNonceURLPatterns(String patterns) {
+        this.noNoncePatterns = patterns;
+
+        if (null != context) {
+            this.noNoncePredicates = createNoNoncePredicates(context, this.noNoncePatterns);
+        }
+    }
+
+    /**
+     * Creates a collection of matchers from a comma-separated string of patterns.
+     *
+     * @param patterns A comma-separated string of URL matching patterns.
+     *
+     * @return A collection of predicates representing the URL patterns.
+     */
+    protected static Collection<Predicate<String>> createNoNoncePredicates(ServletContext context, String patterns) {
+        if (null == patterns || 0 == patterns.trim().length()) {
```

Comment:
why is `#isEmpty()`?

Reply:
In Java 11+ you can use `patterns.isBlank()` which will return true if the string is empty or contains only whitespace

Reply:
Can't be backported

Reply:
@michael-o None of this can be back-ported farther than 10.1.x without significant changes. The whole `jakarta.*` namespace needs to change, `Predicate` needs to be defined locally, etc. I'm okay using `String.isBlank` here if anyone is passionate about it. I generally prefer things to be as close as possible across the branches, so I would err on the side of using `null`-check-plus-zero-length-check until we dump Tomcat 8.5.x.

Reply:
> why is `#isEmpty()`?

I'm not sure I understand this, @michael-o

Reply:
Passionate is a strong word, so I can't say that I'm "passionate" about using String.isBlank(), but OTOH I do like seeing modern Java in the source code where applicable, so I'm a +1 on that one.